app/com/linkedin/drelephant/tony/heuristics/TaskMemoryHeuristic.java

```diff
@@ -0,0 +1,121 @@
+ * the License.
+ */
+package com.linkedin.drelephant.tony.heuristics;
+
+import com.linkedin.drelephant.analysis.Heuristic;
+import com.linkedin.drelephant.analysis.HeuristicResult;
+import com.linkedin.drelephant.analysis.HeuristicResultDetails;
+import com.linkedin.drelephant.analysis.Severity;
+import com.linkedin.drelephant.configurations.heuristic.HeuristicConfigurationData;
+import com.linkedin.drelephant.tony.data.TonyApplicationData;
+import com.linkedin.drelephant.tony.data.TonyTaskData;
+import com.linkedin.drelephant.tony.util.TonyUtils;
+import com.linkedin.drelephant.util.Utils;
+import com.linkedin.tony.Constants;
+import com.linkedin.tony.TonyConfigurationKeys;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.commons.io.FileUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.log4j.Logger;
+
+
+/**
+ * For each type of task (e.g.: worker, ps), this heuristic checks the max memory used across all tasks of that type
+ * and compares against some thresholds. The final severity is the highest severity across all task types.
+ */
+public class TaskMemoryHeuristic implements Heuristic<TonyApplicationData> {
+  private static final Logger _LOGGER = Logger.getLogger(TaskMemoryHeuristic.class);
+  private static final int DEFAULT_CONTAINER_MEMORY_MB = 2048;
+  private static final String CONTAINER_MEMORY_DEFAULT_MB_CONF = "container_memory_default_mb";
+  private static final String TASK_MEMORY_THRESHOLDS_CONF = "task_memory_thresholds";
+
+  private HeuristicConfigurationData _heuristicConfData;
+  private long defaultContainerMemoryBytes = DEFAULT_CONTAINER_MEMORY_MB * FileUtils.ONE_MB;
+
+  // Initialized to default max memory thresholds
+  private double[] maxMemoryLimits = {0.8, 0.7, 0.6, 0.5};
+
+  /**
+   * Constructor for {@link TaskMemoryHeuristic}.
+   * @param heuristicConfData  the configuration for this heuristic
+   */
+  public TaskMemoryHeuristic(HeuristicConfigurationData heuristicConfData) {
+    this._heuristicConfData = heuristicConfData;
+
+    Map<String, String> params = heuristicConfData.getParamMap();
+    // read default container size
+    if (params.containsKey(CONTAINER_MEMORY_DEFAULT_MB_CONF)) {
+      defaultContainerMemoryBytes = Long.parseLong(params.get(CONTAINER_MEMORY_DEFAULT_MB_CONF)) * FileUtils.ONE_MB;
+    }
+    // read max memory thresholds
+    if (params.containsKey(TASK_MEMORY_THRESHOLDS_CONF)) {
+      maxMemoryLimits = Utils.getParam(params.get(TASK_MEMORY_THRESHOLDS_CONF), maxMemoryLimits.length);
+    }
+  }
+
+  @Override
+  public HeuristicResult apply(TonyApplicationData data) {
+    _LOGGER.debug("Applying TaskMemoryHeuristic");
+    Map<String, Map<Integer, TonyTaskData>> taskMap = data.getTaskMap();
+    Configuration conf = data.getConfiguration();
+
+    Set<String> taskTypes = com.linkedin.tony.util.Utils.getAllJobTypes(conf);
```

Comment:
Just out of curiosity, what's the max count for TaskTypes can be?

Reply:
TonY does not impose any limits. In practice, 4 is the most I've seen (worker, ps, chief, evaluator).